### PR TITLE
DOC: update contributing guidelines for black

### LINKED ISSUE #27226: DOC: update contributing guide for black
After https://github.com/pandas-dev/pandas/pull/27076, we should update the contributing guide on how to use black (and pre-commit hook).

We should also put somewhere a short guideline on how to "update" your PR to resolve conflicts.

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes https://github.com/pandas-dev/pandas/issues/27226 
Starting this by copying some content from dask (https://github.com/dask/dask/pull/4983)

Also related to https://github.com/pandas-dev/pandas/issues/23616 (recommending pre-commit hooks)

### CODE REVIEW

File: doc/source/development/contributing.rst [hunk 1]
```diff
@@ -562,23 +562,38 @@ many errors as possible, but it may not correct *all* of them. Thus, it is
...
    git diff upstream/master -u -- "*.py" | flake8 --diff
 
+to auto-format your code. Additionally, many editors have plugins that will
+apply ``black`` as you edit files.
+
+Optionally, you may wish to setup `pre-commit hooks <https://pre-commit.com/>`_
+to automatically run ``black`` and ``flake8`` when you make a git commit. This
+can be done by installing ``pre-commit``::
+
+   pip install pre-commit
```
Comment (jorisvandenbossche) [inline doc/source/development/contributing.rst]: It seems that the conda-forge package is pre_commit (that's a bit an annoying name difference ..)

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -1,4 +1,5 @@
 - [ ] closes #xxxx
 - [ ] tests added / passed
+- [ ] passes `black pandas`
 - [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
```
Comment (jorisvandenbossche) [inline .github/PULL_REQUEST_TEMPLATE.md]: what should we be recommending here? Still keep both flake8 and black? 
The problem with this version of the black command is also that it does the full code base and takes a while

File: pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    - repo: https://github.com/python/black
+      rev: stable
+      hooks:
+          - id: black
+            language_version: python3.7
+    - repo: https://gitlab.com/pycqa/flake8
+      rev: 3.7.7
+      hooks:
+          - id: flake8
+            language: python_venv
```
Comment (jorisvandenbossche) [inline pre-commit-config.yaml]: @TomAugspurger do you know if there is a difference taking flake8 from its own repo or from the pre-commit-hooks repo? 
In https://github.com/pandas-dev/pandas/issues/23616 you use the pre-commit-hooks one, but in a recent gitter chat you pasted the above?

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+repos:
+    - repo: https://github.com/python/black
+      rev: stable
+      hooks:
+          - id: black
+            language_version: python3.7
+    - repo: https://gitlab.com/pycqa/flake8
```
Comment (jorisvandenbossche) [inline .pre-commit-config.yaml]: @TomAugspurger do you know if there is a difference taking flake8 from its own repo or from the pre-commit-hooks repo? 
In https://github.com/pandas-dev/pandas/issues/23616 you use the pre-commit-hooks one, but in a recent gitter chat you pasted the above?

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (jreback) [reply .github/PULL_REQUEST_TEMPLATE.md]: make black is pretty fast currently so i would just have this

File: doc/source/development/contributing.rst [hunk 5]
```diff
@@ -562,23 +562,38 @@ many errors as possible, but it may not correct *all* of them. Thus, it is
...
 
-:ref:`Continuous Integration <contributing.ci>` will run
-the `flake8 <https://pypi.org/project/flake8>`_ tool
-and report any stylistic errors in your code. Therefore, it is helpful before
-submitting code to run the check yourself on the diff::
+:ref:`Continuous Integration <contributing.ci>` will run those tools and
+report any stylistic errors in your code. Therefore, it is helpful before
+submitting code to run the check yourself::
 
+   black pandas
```
Comment (jreback) [inline doc/source/development/contributing.rst]: I think recommening
```make black`` is better here

File: doc/source/development/contributing.rst [hunk 6]
```diff
@@ -562,23 +562,38 @@ many errors as possible, but it may not correct *all* of them. Thus, it is
...
    git diff upstream/master -u -- "*.py" | flake8 --diff
```
Comment (jreback) [inline doc/source/development/contributing.rst]: ```make lint-diff```

File: doc/source/development/contributing.rst (hunk 5 above)
Comment (jorisvandenbossche) [reply doc/source/development/contributing.rst]: Unfortunately `make` does not work on all platforms

**Comment by jreback** [comment]
ok lgtm.

**Review by WillAyd — APPROVED** [review]
lgtm

**Comment by jreback** [comment]
thanks @jorisvandenbossche

**Pull request merged**